### PR TITLE
Change the signature of the getGroup in the ParameterHandler class

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This implements what is discussed in #409

### CODE REVIEW

File: src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/IParametersHandler.h [hunk 1]
```diff
@@ -181,7 +188,7 @@ class IParametersHandler
      * @warning Please implement the specific version of this method in the Derived class. Please
      * check YarpImplementation::getGroup
      */
-    virtual weak_ptr getGroup(const std::string& name) const = 0;
+    virtual const_weak_ptr getGroup(const std::string& name) const = 0;
```
Comment (S-Dafarra) [inline src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/IParametersHandler.h:191]: I was also proposing to add a met

Comment (GiulioRomualdi) [reply src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/IParametersHandler.h:191]: Yes I've to implement that

File: src/Estimators/tests/FloatingBaseEstimators/InvariantEKFBaseEstimatorTest.cpp [hunk 2]
```diff
@@ -34,68 +34,66 @@ double deg2rad(const double& ang)
...
+    priordevGroup->setParameter("r_contact_frame_position", std::vector<double>{1e-3, 1e-3, 1e-3});
+    REQUIRE(handle->setGroup("PriorsStdDev", priordevGroup));
     return true;
 }
 
 TEST_CASE("Invariant EKF Base Estimator")
 {
     std::shared_ptr<StdImplementation> originalHandler = std::make_shared<StdImplementation>();
-    IParametersHandler::shared_ptr parameterHandler = originalHandler;
+    IParametersHandler::shared_ptr parameterHandler = originalHandler->clone();
```
Comment (S-Dafarra) [inline src/Estimators/tests/FloatingBaseEstimators/InvariantEKFBaseEstimatorTest.cpp:96]: Why is this needed?

File: src/IK/tests/QPInverseKinematicsTest.cpp [hunk 3]
```diff
@@ -96,7 +96,10 @@ InverseKinematicsAndTasks createIK(std::shared_ptr<IParametersHandler> handler,
     // prepare the parameters related to the size of the system
     const Eigen::VectorXd kpRegularization = Eigen::VectorXd::Ones(kinDyn->model().getNrOfDOFs());
     const Eigen::VectorXd weightRegularization = kpRegularization;
-    handler->getGroup("REGULARIZATION_TASK").lock()->setParameter("kp", kpRegularization);
+
+    auto regularizationTaskGroup = handler->getGroup("REGULARIZATION_TASK").lock()->clone();
+    regularizationTaskGroup->setParameter("kp", kpRegularization);
+    REQUIRE(handler->setGroup("REGULARIZATION_TASK", regularizationTaskGroup));
```
Comment (S-Dafarra) [inline src/IK/tests/QPInverseKinematicsTest.cpp:102]: I would prefer this change would have not been needed. With the not ``const`` method these modifications should not be necessary, right?

File: src/IK/tests/QPInverseKinematicsTest.cpp [hunk 4]
```diff
@@ -235,9 +238,9 @@ TEST_CASE("QP-IK")
 
             // Set the frame name
             const std::string controlledFrame = model.getFrameName(numberOfJoints);
-            parameterHandler->getGroup("SE3_TASK")
-                .lock()
-                ->setParameter("frame_name", controlledFrame);
+            auto SE3TaskGroup = parameterHandler->getGroup("SE3_TASK").lock()->clone();
+            SE3TaskGroup->setParameter("frame_name", controlledFrame);
+            REQUIRE(parameterHandler->setGroup("SE3_TASK", SE3TaskGroup));
```
Comment (S-Dafarra) [inline src/IK/tests/QPInverseKinematicsTest.cpp:243]: Same as above

File: src/ParametersHandler/TomlImplementation/include/BipedalLocomotion/ParametersHandler/TomlImplementation.h [hunk 5]
```diff
@@ -220,7 +220,15 @@ class TomlImplementation : public IParametersHandler
...
      */
-    weak_ptr getGroup(const std::string& name) const final;
+    const_weak_ptr getGroup(const std::string& name) const final;
+
+    /**
+     * Get a Group from the handler.
+     * @param name name of the group
+     * @return A pointer to IParametersHandler, if the group is not found the weak pointer cannot
+     * be locked
+     */
```
Comment (S-Dafarra) [inline src/ParametersHandler/TomlImplementation/include/BipedalLocomotion/ParametersHandler/TomlImplementation.h:230]: I would suggest adding some details on the fact that this group can be edited afterwards

File: src/ParametersHandler/YarpImplementation/include/BipedalLocomotion/ParametersHandler/YarpImplementation.h [hunk 6]
```diff
@@ -223,14 +223,22 @@ class YarpImplementation : public IParametersHandler
...
-    weak_ptr getGroup(const std::string& name) const final;
+    const_weak_ptr getGroup(const std::string& name) const final;
+
+    /**
+     * Get a Group from the handler.
+     * @param name name of the group
+     * @return A pointer to IParametersHandler, if the group is not found the weak pointer cannot
+     * be locked
+     */
+    weak_ptr getGroup(const std::string& name) final;
```
Comment (S-Dafarra) [inline src/ParametersHandler/YarpImplementation/include/BipedalLocomotion/ParametersHandler/YarpImplementation.h:234]: Same as for the ``TOML`` implementation. I would mention that the group can be edited.

File: src/ParametersHandler/YarpImplementation/include/BipedalLocomotion/ParametersHandler/YarpImplementation.h [hunk 7]
```diff
@@ -223,14 +223,22 @@ class YarpImplementation : public IParametersHandler
...
 
     /**
      * Set a new group on the handler.
      * @param name name of the group
      * @param newGroup shared pointer to the new group
      */
-    bool setGroup(const std::string& name, shared_ptr newGroup) final;
+    bool setGroup(const std::string& name, const_shared_ptr newGroup) final;
```
Comment (S-Dafarra) [inline src/ParametersHandler/YarpImplementation/include/BipedalLocomotion/ParametersHandler/YarpImplementation.h:241]: This modification is missing in the TOML implementation.

Also, I would document that the input group is copied, hence the modifications to the original pointer will not have any effect.

File: src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/IParametersHandler.h [hunk 8]
```diff
@@ -181,7 +188,17 @@ class IParametersHandler
...
+
+    /**
+     * Get a Group from the handler.
+     * @param name name of the group
+     * @return A pointer to IParametersHandler, if the group is not found the weak pointer cannot
+     * be locked
+     * @warning Please implement the specific version of this method in the Derived class. Please
+     * check YarpImplementation::getGroup
+     */
+    virtual weak_ptr getGroup(const std::string& name) = 0;
```
Comment (S-Dafarra) [inline src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/IParametersHandler.h:201]: The same as for the other implementations. I would suggest adding a comment on the fact that the output group can be edited.

File: src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/IParametersHandler.h [hunk 9]
```diff
@@ -191,7 +208,7 @@ class IParametersHandler
      * check YarpImplementation::setGroup
      * @return true/false in case of success/failure
      */
-    virtual bool setGroup(const std::string& name, shared_ptr newGroup) = 0;
+    virtual bool setGroup(const std::string& name, const_shared_ptr newGroup) = 0;
```
Comment (S-Dafarra) [inline src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/IParametersHandler.h:211]: The same for the YARP implementation. I would document that the content of the input pointer is cloned and hence any modification to the input pointer afterward will not have an effect.

File: src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/StdImplementation.h [hunk 10]
```diff
@@ -205,14 +205,22 @@ class StdImplementation : public IParametersHandler
...
-    weak_ptr getGroup(const std::string& name) const final;
+    const_weak_ptr getGroup(const std::string& name) const final;
+
+      /**
+     * Get a Group from the handler.
+     * @param name name of the group
+     * @return A pointer to IParametersHandler, if the group is not found the weak pointer cannot
+     * be locked
+     */
+    weak_ptr getGroup(const std::string& name) final;
```
Comment (S-Dafarra) [inline src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/StdImplementation.h:216]: Same as the other implementations.

File: src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/StdImplementation.h [hunk 11]
```diff
@@ -205,14 +205,22 @@ class StdImplementation : public IParametersHandler
...
 
     /**
      * Set a new group on the handler.
      * @param name name of the group
      * @param newGroup shared pointer to the new group
      */
-    bool setGroup(const std::string& name, shared_ptr newGroup) final;
+    bool setGroup(const std::string& name, const_shared_ptr newGroup) final;
```
Comment (S-Dafarra) [inline src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/StdImplementation.h:223]: Same as the other implementations.

File: src/ParametersHandler/tests/ParametersHandlerTest.cpp [hunk 12]
```diff
@@ -66,13 +66,12 @@ TEST_CASE("Get parameters")
         StdImplementation::shared_ptr newGroup = std::make_shared<StdImplementation>();
         REQUIRE(parameterHandler->setGroup("CARTOONS", newGroup));
         StdImplementation::shared_ptr groupHandler
-            = parameterHandler->getGroup("CARTOONS").lock();
+            = parameterHandler->getGroup("CARTOONS").lock()->clone();
```
Comment (S-Dafarra) [inline src/ParametersHandler/tests/ParametersHandlerTest.cpp:69]: In principle, this modification should not be necessary.

File: src/ParametersHandler/tests/ParametersHandlerTest.cpp [hunk 13]
```diff
@@ -81,7 +80,7 @@ TEST_CASE("Get parameters")
...
```
Comment (S-Dafarra) [inline src/ParametersHandler/tests/ParametersHandlerTest.cpp:83]: Same here, I think this modification is not necessary.

File: src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp [hunk 14]
```diff
@@ -124,12 +124,13 @@ TEST_CASE("Get parameters")
...
     {
-        IParametersHandler::shared_ptr groupHandler = parameterHandler->getGroup("CARTOONS").lock();
-        REQUIRE_FALSE(groupHandler);
+        REQUIRE_FALSE(parameterHandler->getGroup("CARTOONS").lock());
         IParametersHandler::shared_ptr setGroup = std::make_shared<YarpImplementation>();
         REQUIRE(parameterHandler->setGroup("CARTOONS", setGroup));
 
-        groupHandler = parameterHandler->getGroup("CARTOONS").lock(); //now the pointer should be lockable
+        IParametersHandler::shared_ptr groupHandler
+            = parameterHandler->getGroup("CARTOONS").lock()->clone(); // now the pointer should be
```
Comment (S-Dafarra) [inline src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp:132]: Same as above, although it may also make sense to test the ``const`` version of the method. So some of these ``clone`` may remain.

File: src/RobotInterface/tests/DummyImplementation/DummySensorBridgeTest.cpp [hunk 15]
```diff
@@ -5,11 +5,10 @@
  * distributed under the terms of the GNU Lesser General Public License v2.1 or any later version.
  */
 
-
 // Catch2
-#include <catch2/catch.hpp>
 #include <BipedalLocomotion/ParametersHandler/IParametersHandler.h>
 #include <BipedalLocomotion/ParametersHandler/StdImplementation.h>
+#include <catch2/catch.hpp>
```
Comment (S-Dafarra) [inline src/RobotInterface/tests/DummyImplementation/DummySensorBridgeTest.cpp:11]: Also the comment ``// Catch2`` should be moved.

File: src/TSID/tests/QPFixedBaseTSIDTest.cpp [hunk 16]
```diff
@@ -99,8 +99,11 @@ TSIDAndTasks createTSID(std::shared_ptr<IParametersHandler> handler,
     const Eigen::VectorXd kdRegularization = 2 * Eigen::VectorXd::Ones(kinDyn->model().getNrOfDOFs());
     const Eigen::VectorXd weightRegularization = 1 * Eigen::VectorXd::Ones(kinDyn->model().getNrOfDOFs());
 
-    handler->getGroup("REGULARIZATION_TASK").lock()->setParameter("kp", kpRegularization);
-    handler->getGroup("REGULARIZATION_TASK").lock()->setParameter("kd", kdRegularization);
+    auto regularizationTaskGroup = handler->getGroup("REGULARIZATION_TASK").lock()->clone();
+
+    regularizationTaskGroup->setParameter("kp", kpRegularization);
+    regularizationTaskGroup->setParameter("kd", kdRegularization);
+    REQUIRE(handler->setGroup("REGULARIZATION_TASK", regularizationTaskGroup));
```
Comment (S-Dafarra) [inline src/TSID/tests/QPFixedBaseTSIDTest.cpp:106]: These modifications should not be necessary.

File: src/TSID/tests/QPFixedBaseTSIDTest.cpp [hunk 17]
```diff
@@ -250,9 +253,10 @@ TEST_CASE("QP-TSID")
 
             // Set the frame name
             const std::string controlledFrame = model.getFrameName(numberOfJoints);
-            parameterHandler->getGroup("EE_SE3_TASK")
-                .lock()
-                ->setParameter("frame_name", controlledFrame);
+            auto eeSE3TaskGroup = parameterHandler->getGroup("EE_SE3_TASK").lock()->clone();
+            eeSE3TaskGroup->setParameter("frame_name", controlledFrame);
+            REQUIRE(parameterHandler->setGroup("EE_SE3_TASK", eeSE3TaskGroup));
+
```
Comment (S-Dafarra) [inline src/TSID/tests/QPFixedBaseTSIDTest.cpp:259]: Same here

File: utilities/mas-imu-test/include/BipedalLocomotion/.#MasImuTest.h [hunk 18]
```diff
@@ -0,0 +1 @@
+gromualdi@iiticublap261.23196:1630657517
```
Comment (S-Dafarra) [inline utilities/mas-imu-test/include/BipedalLocomotion/.#MasImuTest.h:1]: This file should be removed. Consider updating the ``.gitignore``.

File: utilities/mas-imu-test/src/.#MasImuTest.cpp [hunk 19]
```diff
@@ -0,0 +1 @@
+gromualdi@iiticublap261.23196:1630657517
```
Comment (S-Dafarra) [inline utilities/mas-imu-test/src/.#MasImuTest.cpp:1]: Same here

File: utilities/mas-imu-test/src/MasImuTest.cpp [hunk 20]
```diff
@@ -556,11 +556,11 @@ double MasImuTest::MasImuData::maxVariation()
     return maxVariation;
 }
 
-bool MasImuTest::MasImuData::setup(ParametersHandler::YarpImplementation::shared_ptr group,
+bool MasImuTest::MasImuData::setup(ParametersHandler::YarpImplementation::const_shared_ptr group,
                                    std::shared_ptr<CommonData> commonDataPtr)
 {
     m_commonDataPtr = commonDataPtr;
-    m_group = group;
+    m_group = group->clone();
```
Comment (S-Dafarra) [inline utilities/mas-imu-test/src/MasImuTest.cpp:563]: You can simply change the signature of ``m_group``.

**Comment by traversaro** [comment]
I removed myself from the review, feel free to add me back if this PR is ready.

**Comment by GiulioRomualdi** [comment]
This feature is no needed anymore and the code changed a lot during the years. Closing
